File: go/multipaxos/network/tcp.go
package network

import (
	"bufio"
	"encoding/json"
	"net"
	"net/textproto"
	"sync"
)

type Message struct {
	Type      uint8
	ChannelId uint64
	Msg       string
}

type ChannelMap struct {
	sync.Mutex
	Channels map[uint64]chan string
}

func handleOutgoingRequests(stream net.Conn, requestChan chan string) {
	for request := range requestChan {
		request = request + "\n"
		requestBuffer := []byte(request)
		_, err := stream.Write(requestBuffer)
		if err != nil {
			return
		}
	}
}

func handleIncomingResponses(stream net.Conn, channels *ChannelMap) {
	reader := textproto.NewReader(bufio.NewReader(stream))
	for {
		line, err := reader.ReadLineBytes()
		if err != nil {
			break
		}
		var response Message
		err = json.Unmarshal(line, &response)
		channels.Lock()
		if responseChan, ok := channels.Channels[response.ChannelId]; ok {
			responseChan <- response.Msg
		}
		channels.Unlock()
	}
	channels.Lock()
	defer channels.Unlock()
	for key := range channels.Channels {
		delete(channels.Channels, key)
	}
}

type TcpLink struct {
	requestChan chan string
}

func NewTcpLink(addr string, channels *ChannelMap) *TcpLink {
	var stream net.Conn
	for {
		var err error
		stream, err = net.Dial("tcp", addr)
		if err == nil {
			break
		}
	}
	go handleIncomingResponses(stream, channels)
	requestChan := make(chan string)
	go handleOutgoingRequests(stream, requestChan)
	return &TcpLink{
		requestChan: requestChan,
	}
}

func (t *TcpLink) Start() {

}

func (t *TcpLink) SendAwaitResponse(msgType MessageType, channelId uint64, msg string) {
	request, _ := json.Marshal(Message{
		Type:      uint8(msgType),
		ChannelId: channelId,
		Msg:       msg,
	})
	t.requestChan <- string(request)
}

File: rust-tcp/replicant/src/replicant/multipaxos/tcp.rs
use std::collections::HashMap;
use std::sync::Arc;
use serde::{Deserialize, Serialize};
use tokio::io::{AsyncBufReadExt, AsyncWriteExt, BufReader};
use tokio::net::tcp::{OwnedReadHalf, OwnedWriteHalf};
use tokio::net::TcpStream;
use tokio::sync::mpsc;
use tokio::sync::mpsc::{Sender, UnboundedReceiver, UnboundedSender};
use crate::replicant::multipaxos::msg::MessageType;

#[derive(Serialize, Deserialize)]
pub struct Message {
    pub r#type: u8,
    pub channel_id: u64,
    pub msg: String,
}

async fn handle_outgoing_requests(
    mut write_half: OwnedWriteHalf,
    mut request_recv: UnboundedReceiver<String>
) {
    while let Some(mut request) = request_recv.recv().await {
        request.push('\n');
        write_half.write_all(request.as_bytes()).await
            .expect("peer cannot send msg");
    }
}

async fn handle_incoming_responses(
    read_half: OwnedReadHalf,
    channels: Arc<tokio::sync::Mutex<HashMap<u64, Sender<String>>>>
) {
    let mut read_half = BufReader::new(read_half);
    let mut line = String::new();
    while let Ok(n) = read_half.read_line(&mut line).await {
        if n == 0 {
            break;
        }
        let request: Message = match serde_json::from_str(&line) {
            Ok(request) => request,
            Err(_) => continue
        };
        let channels = channels.lock().await;
        let response_send = channels.get(&request.channel_id);
        if let Some(response_send) = response_send {
            response_send.send(request.msg).await;
        }
        drop(channels);
        line.clear();
    }
    let mut channels = channels.lock().await;
    channels.clear();
}

pub struct TcpLink {
    request_sender: UnboundedSender<String>,
}

impl TcpLink {
    pub async fn new(
        addr: &str,
        channels: Arc<tokio::sync::Mutex<HashMap<u64, Sender<String>>>>
    ) -> Self {
        let stream;
        loop {
            match TcpStream::connect(addr).await {
                Ok(s) => {stream = s; stream.set_nodelay(true); break;},
                _ => (),
            }
        }
        let (read_half, write_half) = stream.into_split();
        tokio::spawn(async move {
            handle_incoming_responses(read_half, channels).await;
        });
        let (request_sender, request_recv) = mpsc::unbounded_channel();
        tokio::spawn(async move {
            handle_outgoing_requests(write_half, request_recv).await;
        });
        Self {
            request_sender,
        }
    }

    pub async fn send_await_response(
        &self,
        msg_type: MessageType,
        channel_id: u64,
        msg: &str)
    {
        let tcp_request = serde_json::to_string(&Message{
            r#type: msg_type as u8,
            channel_id,
            msg: msg.to_string(),
        }).unwrap();
        self.request_sender.send(tcp_request).
            expect("cannot send request via channel");
    }
}
